File: libs/common/src/typegraph.rs
// Copyright Metatype OÜ under the Elastic License 2.0 (ELv2). See LICENSE.md for usage.

use std::path::PathBuf;

use anyhow::{bail, Result};
use indexmap::IndexMap;
#[cfg(feature = "codegen")]
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use serde_with::skip_serializing_none;

#[cfg_attr(feature = "codegen", derive(JsonSchema))]
#[derive(Serialize, Deserialize, Debug)]
pub struct Typegraph {
    #[serde(rename = "$id")]
    pub id: String,
    pub types: Vec<TypeNode>,
    pub materializers: Vec<Materializer>,
    pub runtimes: Vec<TGRuntime>,
    pub policies: Vec<Policy>,
    pub meta: TypeMeta,
    #[serde(skip)]
    pub path: Option<PathBuf>,
    #[serde(skip)]
    pub deps: Vec<PathBuf>,
}

#[cfg_attr(feature = "codegen", derive(JsonSchema))]
#[derive(Serialize, Deserialize, Debug)]
pub struct Cors {
    pub allow_origin: Vec<String>,
    pub allow_headers: Vec<String>,
    pub expose_headers: Vec<String>,
    #[serde(default)]
    pub allow_methods: Vec<String>,
    pub allow_credentials: bool,
    pub max_age_sec: Option<u32>,
}

#[cfg_attr(feature = "codegen", derive(JsonSchema))]
#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "lowercase")]
pub enum AuthProtocol {
    OAuth2,
    Jwt,
    Basic,
}

#[cfg_attr(feature = "codegen", derive(JsonSchema))]
#[derive(Serialize, Deserialize, Debug)]
pub struct Auth {
    pub name: String,
    pub protocol: AuthProtocol,
    pub auth_data: IndexMap<String, Value>,
}

#[cfg_attr(feature = "codegen", derive(JsonSchema))]
#[derive(Serialize, Deserialize, Debug)]
pub struct Rate {
    pub window_limit: u32,
    pub window_sec: u32,
    pub query_limit: u32,
    pub context_identifier: Option<String>,
    pub local_excess: u32,
}

#[cfg_attr(feature = "codegen", derive(JsonSchema))]
#[derive(Serialize, Deserialize, Debug)]
pub struct TypeMeta {
    pub secrets: Vec<String>,
    pub cors: Cors,
    pub auths: Vec<Auth>,
    pub rate: Option<Rate>,
    pub version: String,
}

#[cfg_attr(feature = "codegen", derive(JsonSchema))]
#[skip_serializing_none]
#[derive(Serialize, Deserialize, Debug)]
pub struct TypeNodeBase {
    pub title: String,
    pub runtime: u32,
    pub policies: Vec<PolicyIndices>,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub injection: Option<String>,
    #[serde(default, skip_serializing_if = "serde_json::Value::is_null")]
    pub inject: Value,
    #[serde(default, rename = "enum")]
    pub enumeration: Option<Vec<Value>>,
    #[serde(default)]
    pub config: IndexMap<String, serde_json::Value>,
}

#[cfg_attr(feature = "codegen", derive(JsonSchema))]
#[skip_serializing_none]
#[derive(Serialize, Deserialize, Debug)]
#[serde(tag = "type", rename_all = "lowercase")]
pub enum TypeNode {
    Optional {
        #[serde(flatten)]
        base: TypeNodeBase,
        item: u32,
        #[serialize_always]
        default_value: Option<serde_json::Value>,
    },
    Boolean {
        #[serde(flatten)]
        base: TypeNodeBase,
    },
    #[serde(rename_all = "camelCase")]
    Number {
        #[serde(flatten)]
        base: TypeNodeBase,
        minimum: Option<f64>,
        maximum: Option<f64>,
        exclusive_minimum: Option<f64>,
        exclusive_maximum: Option<f64>,
        multiple_of: Option<f64>,
    },
    #[serde(rename_all = "camelCase")]
    Integer {
        #[serde(flatten)]
        base: TypeNodeBase,
        minimum: Option<i64>,
        maximum: Option<i64>,
        exclusive_minimum: Option<i64>,
        exclusive_maximum: Option<i64>,
        multiple_of: Option<i64>,
    },
    #[serde(rename_all = "camelCase")]
    String {
        #[serde(flatten)]
        base: TypeNodeBase,
        min_length: Option<i64>,
        max_length: Option<i64>,
        pattern: Option<String>,
        format: Option<String>,
    },
    Object {
        #[serde(flatten)]
        base: TypeNodeBase,
        properties: IndexMap<String, u32>,
        #[serde(default)]
        required: Vec<String>,
    },
    #[serde(rename_all = "camelCase")]
    Array {
        #[serde(flatten)]
        base: TypeNodeBase,
        items: u32,
        max_items: Option<u32>,
        min_items: Option<u32>,
        unique_items: Option<bool>,
    },
    Function {
        #[serde(flatten)]
        base: TypeNodeBase,
        input: u32,
        output: u32,
        materializer: u32,
        #[serialize_always]
        rate_weight: Option<u32>,
        rate_calls: bool,
    },
    #[serde(rename_all = "camelCase")]
    Union {
        #[serde(flatten)]
        base: TypeNodeBase,
        /// Array of indexes of the nodes that are used as subschemes in the
        /// anyOf field of JSON Schema.
        any_of: Vec<u32>,
    },
    #[serde(rename_all = "camelCase")]
    Either {
        #[serde(flatten)]
        base: TypeNodeBase,
        /// Array of indexes of the nodes that are used as subschemes in the
        /// oneOf field of JSON Schema.
        one_of: Vec<u32>,
    },
    Any {
        #[serde(flatten)]
        base: TypeNodeBase,
    },
}

impl TypeNode {
    pub fn base(&self) -> &TypeNodeBase {
        use TypeNode::*;
        match self {
            Optional { base, .. }
            | Boolean { base, .. }
            | Number { base, .. }
            | Integer { base, .. }
            | String { base, .. }
            | Object { base, .. }
            | Array { base, .. }
            | Function { base, .. }
            | Union { base, .. }
            | Either { base, .. }
            | Any { base, .. } => base,
        }
    }
}

#[cfg_attr(feature = "codegen", derive(JsonSchema))]
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "lowercase")]
pub enum EffectType {
    Create,
    Update,
    Upsert,
    Delete,
}

#[cfg_attr(feature = "codegen", derive(JsonSchema))]
#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Effect {
    effect: Option<EffectType>,
    idempotent: bool,
}

#[cfg_attr(feature = "codegen", derive(JsonSchema))]
#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Materializer {
    pub name: String,
    pub runtime: u32,
    pub effect: Effect,
    pub data: IndexMap<String, Value>,
}

#[cfg_attr(feature = "codegen", derive(JsonSchema))]
#[derive(Serialize, Deserialize, Debug)]
pub struct TGRuntime {
    pub name: String,
    pub data: IndexMap<String, Value>,
}

#[cfg_attr(feature = "codegen", derive(JsonSchema))]
#[derive(Serialize, Deserialize, Debug)]
pub struct Policy {
    pub name: String,
    pub materializer: u32,
}

#[cfg_attr(feature = "codegen", derive(JsonSchema))]
#[skip_serializing_none]
#[derive(Serialize, Deserialize, Debug)]
pub struct PolicyIndicesByEffect {
    pub none: Option<u32>,
    pub create: Option<u32>,
    pub delete: Option<u32>,
    pub update: Option<u32>,
    pub upsert: Option<u32>,
}

#[cfg_attr(feature = "codegen", derive(JsonSchema))]
#[skip_serializing_none]
#[derive(Serialize, Deserialize, Debug)]
#[serde(untagged)]
pub enum PolicyIndices {
    Policy(u32),
    EffectPolicies(PolicyIndicesByEffect),
}

impl Typegraph {
    pub fn name(&self) -> Result<String> {
        let root_type = &self.types[0];
        match root_type {
            TypeNode::Object { base, .. } => Ok(base.title.clone()),
            _ => bail!("invalid variant for root type"),
        }
    }
}

impl TypeNode {
    pub fn get_struct_fields(&self) -> Result<IndexMap<String, u32>> {
        if let TypeNode::Object { properties, .. } = &self {
            Ok(properties.clone())
        } else {
            bail!("node is not an object variant, found: {self:#?}")
        }
    }
}

#[cfg_attr(feature = "codegen", derive(JsonSchema))]
#[derive(Serialize, Deserialize)]
pub struct FunctionMatData {
    pub script: String,
}

#[cfg_attr(feature = "codegen", derive(JsonSchema))]
#[derive(Debug, Serialize, Deserialize)]
pub struct ModuleMatData {
    pub code: String,
}

#[cfg_attr(feature = "codegen", derive(JsonSchema))]
#[skip_serializing_none]
#[derive(Serialize, Deserialize, Debug)]
pub struct PrismaRuntimeData {
    pub name: String,
    pub datamodel: String,
    pub connection_string_secret: String,
    pub models: Vec<u32>,
    pub migrations: Option<String>,
    #[serde(default)]
    pub create_migration: bool,
}

File: typegate/src/types/typegraph.ts
// Copyright Metatype OÜ, licensed under the Elastic License 2.0.
// SPDX-License-Identifier: Elastic-2.0

// deno-lint-ignore-file no-explicit-any no-empty-interface

export type OptionalNode = {
  type: "optional";
  title: string;
  runtime: number;
  policies: PolicyIndices[];
  description?: string | null;
  injection?: Injection | null;
  enum?: string[] | null;
  config?: {
    [k: string]: unknown;
  };
  as_id: boolean;
  item: number;
  default_value?: any;
};
export type BooleanNode = {
  type: "boolean";
  title: string;
  runtime: number;
  policies: PolicyIndices[];
  description?: string | null;
  injection?: Injection | null;
  enum?: string[] | null;
  config?: {
    [k: string]: unknown;
  };
  as_id: boolean;
};
export type FloatNode = {
  type: "float";
  title: string;
  runtime: number;
  policies: PolicyIndices[];
  description?: string | null;
  injection?: Injection | null;
  enum?: string[] | null;
  config?: {
    [k: string]: unknown;
  };
  as_id: boolean;
  minimum?: number | null;
  maximum?: number | null;
  exclusiveMinimum?: number | null;
  exclusiveMaximum?: number | null;
  multipleOf?: number | null;
};
export type IntegerNode = {
  type: "integer";
  title: string;
  runtime: number;
  policies: PolicyIndices[];
  description?: string | null;
  injection?: Injection | null;
  enum?: string[] | null;
  config?: {
    [k: string]: unknown;
  };
  as_id: boolean;
  minimum?: number | null;
  maximum?: number | null;
  exclusiveMinimum?: number | null;
  exclusiveMaximum?: number | null;
  multipleOf?: number | null;
};
export type StringNode = {
  type: "string";
  title: string;
  runtime: number;
  policies: PolicyIndices[];
  description?: string | null;
  injection?: Injection | null;
  enum?: string[] | null;
  config?: {
    [k: string]: unknown;
  };
  as_id: boolean;
  minLength?: number | null;
  maxLength?: number | null;
  pattern?: string | null;
  format?: StringFormat | null;
};
export type FileNode = {
  type: "file";
  title: string;
  runtime: number;
  policies: PolicyIndices[];
  description?: string | null;
  injection?: Injection | null;
  enum?: string[] | null;
  config?: {
    [k: string]: unknown;
  };
  as_id: boolean;
  minSize?: number | null;
  maxSize?: number | null;
  mimeTypes?: string[] | null;
};
export type ObjectNode = {
  type: "object";
  title: string;
  runtime: number;
  policies: PolicyIndices[];
  description?: string | null;
  injection?: Injection | null;
  enum?: string[] | null;
  config?: {
    [k: string]: unknown;
  };
  as_id: boolean;
  properties: {
    [k: string]: number;
  };
  required?: string[];
};
export type ArrayNode = {
  type: "array";
  title: string;
  runtime: number;
  policies: PolicyIndices[];
  description?: string | null;
  injection?: Injection | null;
  enum?: string[] | null;
  config?: {
    [k: string]: unknown;
  };
  as_id: boolean;
  items: number;
  maxItems?: number | null;
  minItems?: number | null;
  uniqueItems?: boolean | null;
};
export type FunctionNode = {
  type: "function";
  title: string;
  runtime: number;
  policies: PolicyIndices[];
  description?: string | null;
  injection?: Injection | null;
  enum?: string[] | null;
  config?: {
    [k: string]: unknown;
  };
  as_id: boolean;
  input: number;
  output: number;
  materializer: number;
  rate_weight?: number | null;
  rate_calls: boolean;
};
export type UnionNode = {
  type: "union";
  title: string;
  runtime: number;
  policies: PolicyIndices[];
  description?: string | null;
  injection?: Injection | null;
  enum?: string[] | null;
  config?: {
    [k: string]: unknown;
  };
  as_id: boolean;
  anyOf: number[];
};
export type EitherNode = {
  type: "either";
  title: string;
  runtime: number;
  policies: PolicyIndices[];
  description?: string | null;
  injection?: Injection | null;
  enum?: string[] | null;
  config?: {
    [k: string]: unknown;
  };
  as_id: boolean;
  oneOf: number[];
};
export type AnyNode = {
  type: "any";
  title: string;
  runtime: number;
  policies: PolicyIndices[];
  description?: string | null;
  injection?: Injection | null;
  enum?: string[] | null;
  config?: {
    [k: string]: unknown;
  };
  as_id: boolean;
};
export type TypeNode =
  | OptionalNode
  | BooleanNode
  | FloatNode
  | IntegerNode
  | StringNode
  | FileNode
  | ObjectNode
  | ArrayNode
  | FunctionNode
  | UnionNode
  | EitherNode
  | AnyNode;
export type PolicyIndices = number | PolicyIndicesByEffect;
export type Injection = {
  source: "static";
  data: InjectionDataFor_String;
} | {
  source: "context";
  data: InjectionDataFor_String;
} | {
  source: "secret";
  data: InjectionDataFor_String;
} | {
  source: "parent";
  data: InjectionDataForUint32;
} | {
  source: "dynamic";
  data: InjectionDataFor_String;
};
export type InjectionDataFor_String = SingleValueFor_String | {
  [k: string]: string;
};
export type InjectionDataForUint32 = SingleValueForUint32 | {
  [k: string]: number;
};
export type StringFormat =
  | "uuid"
  | "email"
  | "uri"
  | "json"
  | "hostname"
  | "ean"
  | "date"
  | "date-time"
  | "phone";
export type EffectType = "create" | "update" | "delete" | "none";
export type TGRuntime = KnownRuntime | UnknownRuntime;
export type KnownRuntime = {
  name: "deno";
  data: DenoRuntimeData;
} | {
  name: "graphql";
  data: GraphQLRuntimeData;
} | {
  name: "http";
  data: HTTPRuntimeData;
} | {
  name: "python_wasi";
  data: PythonRuntimeData;
} | {
  name: "random";
  data: RandomRuntimeData;
} | {
  name: "prisma";
  data: PrismaRuntimeData;
} | {
  name: "prisma_migration";
  data: PrismaMigrationRuntimeData;
} | {
  name: "s3";
  data: S3RuntimeData;
} | {
  name: "temporal";
  data: TemporalRuntimeData;
} | {
  name: "wasmedge";
  data: WasmEdgeRuntimeData;
} | {
  name: "typegate";
  data: TypegateRuntimeData;
} | {
  name: "typegraph";
  data: TypegraphRuntimeData;
};
export type Cardinality = "optional" | "one" | "many";
export type AuthProtocol = "oauth2" | "jwt" | "basic";
export type S3Materializer = {
  name: "presign_get";
  data: {
    bucket: string;
    expiry_secs?: number | null;
  };
} | {
  name: "presign_put";
  data: {
    bucket: string;
    content_type?: string | null;
    expiry_secs?: number | null;
  };
} | {
  name: "list";
  data: {
    bucket: string;
  };
} | {
  name: "upload";
  data: {
    bucket: string;
  };
} | {
  name: "upload_all";
  data: {
    bucket: string;
  };
};
export interface Typegraph {
  $id: string;
  types: TypeNode[];
  materializers: Materializer[];
  runtimes: TGRuntime[];
  policies: Policy[];
  meta: TypeMeta;
}
export interface PolicyIndicesByEffect {
  none?: number | null;
  create?: number | null;
  delete?: number | null;
  update?: number | null;
}
export interface SingleValueFor_String {
  value: string;
}
export interface SingleValueForUint32 {
  value: number;
}
export interface Materializer {
  name: string;
  runtime: number;
  effect: Effect;
  data: {
    [k: string]: unknown;
  };
}
export interface Effect {
  effect?: EffectType | null;
  idempotent: boolean;
}
export interface DenoRuntimeData {
  worker: string;
  permissions: {
    [k: string]: unknown;
  };
}
export interface GraphQLRuntimeData {
  endpoint: string;
}
export interface HTTPRuntimeData {
  endpoint: string;
  cert_secret?: string | null;
  basic_auth_secret?: string | null;
}
export interface PythonRuntimeData {
  config?: string | null;
}
export interface RandomRuntimeData {
  seed: number;
  reset?: string | null;
}
export interface PrismaRuntimeData {
  name: string;
  connection_string_secret: string;
  models: number[];
  relationships: Relationship[];
  migration_options?: MigrationOptions | null;
}
export interface Relationship {
  name: string;
  left: RelationshipModel;
  right: RelationshipModel;
}
export interface RelationshipModel {
  type_idx: number;
  field: string;
  cardinality: Cardinality;
}
export interface MigrationOptions {
  migration_files?: string | null;
  create: boolean;
  reset: boolean;
}
export interface PrismaMigrationRuntimeData {
}
export interface S3RuntimeData {
  host_secret: string;
  region_secret: string;
  access_key_secret: string;
  secret_key_secret: string;
  path_style_secret: string;
}
export interface TemporalRuntimeData {
  name: string;
  host: string;
}
export interface WasmEdgeRuntimeData {
  config?: string | null;
}
export interface TypegateRuntimeData {
}
export interface TypegraphRuntimeData {
}
export interface UnknownRuntime {
  name: string;
  data: {
    [k: string]: unknown;
  };
}
export interface Policy {
  name: string;
  materializer: number;
}
export interface TypeMeta {
  prefix?: string | null;
  secrets: string[];
  queries: Queries;
  cors: Cors;
  auths: Auth[];
  rate?: Rate | null;
  version: string;
}
export interface Queries {
  dynamic: boolean;
  endpoints: string[];
}
export interface Cors {
  allow_origin: string[];
  allow_headers: string[];
  expose_headers: string[];
  allow_methods?: string[];
  allow_credentials: boolean;
  max_age_sec?: number | null;
}
export interface Auth {
  name: string;
  protocol: AuthProtocol;
  auth_data: {
    [k: string]: unknown;
  };
}
export interface Rate {
  window_limit: number;
  window_sec: number;
  query_limit: number;
  context_identifier?: string | null;
  local_excess: number;
}
export interface FunctionMatData {
  script: string;
}
export interface ModuleMatData {
  code: string;
}
export interface PrismaOperationMatData {
  table: string;
  operation: string;
}
